### PR TITLE
New Admin Ticket System

### PR DESCRIPTION
**Pull request by Gandalf2k15**
This implements a new admin help system. It utilises a menu and several neat features. Port from Yogstation. Pictures:
**As admin:**
**Main panel**
image
image

**Secondary Panel**
image

**As NON admin:**
image
image

**DEMO**
https://dl.dropboxusercontent.com/u/109295987/ShareX/2016/06/2016-06-15_02-58-56.mp4

INSTRUCTIONS FOR ADMINS:
TO VIEW ALL TICKETS CLICK **Adminlisttickets** IN ADMIN TAB. THIS WILL BRING UP THE PRIMARY CONTROL MENU.
image
Have fun. As per nebyabs request.

### CODE REVIEW

**Comment by Nebyab** [comment]
oh god it still says tickets

**Comment by Gandalf2k15** [comment]
WELL IF YOU THINK ABOUT IT ITS BASICALLY A TICKET FOR HELP

**Comment by Spacedong** [comment]
help

**Comment by Nebyab** [comment]
okay, that doesn't make it not retarded, nor does it make it pass travis

**Comment by Gandalf2k15** [comment]
yeah i fucked it in the ass wheps

**Comment by Gandalf2k15** [comment]
Okay it says adminhelp instead of Tickets

**Comment by Gandalf2k15** [comment]
I'll fix the very obvious typo in a bit

**Comment by Badmex** [comment]
Yes, gj Gandalf

**Comment by doobiedoo23** [comment]
WE HEAVY RP NOW NIGGA

**Comment by Gandalf2k15** [comment]
Nah its for admins, makes their job easier.

**Comment by ghost** [comment]
Let's run this shit +1

**Comment by Gandalf2k15** [comment]
Yep we should give it a day test run as well to ensure its good for us.

**Comment by calibraptor** [comment]
Ehhhh I'm really iffy about this

**Comment by Gandalf2k15** [comment]
Why

**Comment by calibraptor** [comment]
Because it makes it so you have to open up a menu in order to handle adminhelps, instead of being able to just click on the ahelp as it pops up to handle it.

Having something that keeps track of adminhelps is neat but not when it removes valuable functionality.

**Comment by dreamcrab** [comment]
this is probably gonna fuck with how nalarbot handles the admin-watch channel on slack too

**Comment by calibraptor** [comment]
Yeah -1

Why fix what isn't broken?

**Comment by Gandalf2k15** [comment]
Because it was decided upon in the meeting.

**Comment by Gandalf2k15** [comment]
And you can still click the chat to respond.
No one said anything about fixing. Its recreating.

**Comment by Gandalf2k15** [comment]
Already has two +1's. I too would like this, I've tested it along with several other people. IIRC should be fine. It still uses the send2irc proc. Just a different message.

**Comment by Nebyab** [comment]
ok

File: code/__HELPERS/mobs.dm [hunk 1]
```diff
@@ -11,7 +11,189 @@
 		if("amber")		return "fc0"
 		if("albino")	return pick("c","d","e","f") + pick("0","1","2","3","4","5","6","7","8","9") + pick("0","1","2","3","4","5","6","7","8","9")
 		else			return "000"
+/////////////////ADMIN TICKET SYSTEM///////////////////
+/proc/is_admin(var/user)
```
Comment (JohnGinnane) [inline code/__HELPERS/mobs.dm:15]: Should this proc return 1 if user is an admin?

Comment (Gandalf2k15) [inline code/__HELPERS/mobs.dm:15]: ya

File: code/__HELPERS/mobs.dm [hunk 2]
```diff
@@ -11,7 +11,189 @@
...
+
+	var/list/adminhelp_ignored_words = list("unknown","the","a","an","of","monkey","alien","as", "i")
+	var/list/msglist = text2list(msg, " ")
+
+	//generate keywords lookup
+	var/list/surnames = list()
+	var/list/forenames = list()
+	var/list/ckeys = list()
+	for(var/mob/M in mob_list)
+		if(!M.mind && !M.client)
```
Comment (JohnGinnane) [inline code/__HELPERS/mobs.dm:50]: Should this be if they have no mind OR they're not a client?

File: code/modules/admin/tickets/admin_ticket.dm [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+	var/text_admin
+	var/for_admins
+
+/datum/ticket_log/New(var/datum/admin_ticket/parent, var/user, var/text, var/for_admins = 0)
+	src.gametime = gameTimestamp()
+	src.parent = parent
+	if(istype(user, /client))
+		src.user_admin = is_admin(user)
+	else
+		src.user_admin = text
```
Comment (JohnGinnane) [inline code/modules/admin/tickets/admin_ticket.dm:21]: Why set user_admin to text?

File: code/__HELPERS/mobs.dm (hunk 2 above)
Comment (Gandalf2k15) [inline code/__HELPERS/mobs.dm:50]: Thats how yog done it.

File: code/modules/admin/tickets/admin_ticket.dm (hunk 3 above)
Comment (Gandalf2k15) [inline code/modules/admin/tickets/admin_ticket.dm:21]: I dont know.

**Comment by Sarcalogo** [comment]
What's with 

 code/modules/awaymissions/awaymissionconfig.txt 

being changed?

**Comment by nalarmg** [comment]
I'm against some sections of this particular "recreation"/port. 

I agree that a way of storing and tracking is needed, but I feel anything more is unsuitable for the pace the server works with. This sort of system seems designed for slower paced servers and rounds, and I'm not sure that the way this is being handled will be beneficial in the long run.

**Comment by Sarcalogo** [comment]
No, it's important for the PLAYER's to understand. Okay, it says my admin help was accepted so I know someone looked at it. Oh, it was resolved, I guess that means the guy was banned or wasn't an antag. 

It's much more useful for the players, who often hear NOTHING back from the admins than it would be for the admins.

**Comment by Gandalf2k15** [comment]
We need a testrun

**Comment by Gandalf2k15** [comment]
I don't actually know about the config. Its not supposed to be in here, removing.

File: code/modules/admin/tickets/admin_ticket.dm (hunk 3 above)
Comment (TDcoolguy300) [inline code/modules/admin/tickets/admin_ticket.dm:21]: Based on this line, it's a fallback should the istype fail above, such as the client suddenly going missing midprocess. Should be fine.

**Comment by TDcoolguy300** [comment]
Got approved for a test run, Blame Mcbawbags and Gandalf should this break shit. Already checked it and it should work just fine.

**Pull request merged**